src/main/java/org/graylog/integrations/aws/service/CloudWatchService.java

```diff
@@ -54,4 +60,32 @@ public LogGroupsResponse getLogGroupNames(String region, String awsAccessKeyId,
         LOG.debug("Log groups queried: [{}]", groupNameList);
         return LogGroupsResponse.create(groupNameList, groupNameList.size());
     }
+
+    public String addSubscriptionFilter(CloudWatchLogsClient cloudWatch, String logGroup, String streamArn,
+                                        String roleArn, String filterName,
+                                        String filterPattern) throws InterruptedException {
+        PutSubscriptionFilterRequest putSubscriptionFilterRequest =
+                PutSubscriptionFilterRequest.builder()
+                                            .logGroupName(logGroup)
+                                            .filterName(filterName)
+                                            .filterPattern(filterPattern)
+                                            .destinationArn(streamArn)
+                                            .roleArn(roleArn)
+                                            .distribution(Distribution.BY_LOG_STREAM)
+                                            .build();
+        sleep(2_000);
+        try {
+            cloudWatch.putSubscriptionFilter(putSubscriptionFilterRequest);
+            final String responseMessage = String.format("Success. The subscription filter [%s] was added to [%s].",
+                                                         filterName, logGroup);
+            return responseMessage;
+        } catch (Exception e) {
+            final String specificError = ExceptionUtils.formatMessageCause(e);
+            final String responseMessage = String.format("Attempt to add subscription [%s] to Cloudwatch log group " +
+                                                         "[%s] failed due to the following exception: [%s]",
+                                                         filterName, logGroup, specificError);
+            LOG.error(responseMessage);
+            throw new InternalServerErrorException(responseMessage, e);
```

Comment:
What do you think about throwing `javax.ws.rs.BadRequestException` here instead? `BadRequestException` will respond with a 400 HTTP code indicating that something is wrong with the request parameters. 

`InternalServerErrorException` returns a 500 HTTP code which generally says that it was the Graylog servers fault. 

I think the `BadRequestException` fits better, but happy to discuss more too. 

More info on statuses: https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.1

Reply:
Good thinking. That exception makes more sense.